### ISSUE TITLE
[numpy] Add `iscomplexobj` and `isrealobj`

### ISSUE THREAD

**Issue by kshitij12345** [pytorch/pytorch#65306]
Pytorch currently has `torch.is_complex` which takes a Tensor and returns whether the dtype is complex for not. (`np.iscomplexobj` is the corresponding similar function in NumPy)

NumPy has a function `iscomplex` which takes an array and returns a boolean array where value is True is corresponding element is complex.

To avoid confusion for the users coming from NumPy, it would be nice if PyTorch adds `iscomplexobj` (alias to already existing `is_complex`) and also `isrealobj`.

Documentation:
torch.is_complex : https://pytorch.org/docs/stable/generated/torch.is_complex.html
np.iscomplex : https://numpy.org/doc/stable/reference/generated/numpy.iscomplex.html#numpy.iscomplex
np.iscomplexobj : https://numpy.org/doc/stable/reference/generated/numpy.iscomplexobj.html#numpy.iscomplexobj
np.isrealobj : https://numpy.org/doc/stable/reference/generated/numpy.isrealobj.html#numpy.isrealobj

**Comment by kshitij12345** [comment]
@AnirudhDagar assigning the issue as per our offline conversation. Let's wait till there is a clear direction of what is to be pursued.

**Comment by anjali411** [comment]
This makes sense. However, we have a history of using is_<dtype> in PyTorch so making this change might be a bit confusing for users

**Comment by mruberry** [comment]
> This makes sense. However, we have a history of using is_ in PyTorch so making this change might be a bit confusing for users (as you pointed out). I like the idea of adding functions similar to numpy and keeping is_complex too.

We can be clear that iscomplexobj is an alias to is_complex and in the future we should be more diligent about noting which signatures are added for NumPy/SciPy/libroas/etc. compat, too. 

`iscomplexobj` is a weird name so it will be hard to mistake for something else, but looking ahead `np.iscomplex` is not the same function as `torch.is_complex`, which is a pain. It would be challenging to have both `torch.iscomplex` and `torch.is_complex` and have them be different.

**Comment by rgommers** [comment]
> It would be challenging to have both `torch.iscomplex` and `torch.is_complex` and have them be different.

I agree with that. The NumPy APIs for dtype checks are a bit of a mess, and should be cleaned up at some point - I opened an issue for that last year: https://github.com/numpy/numpy/issues/17325. Adding aliases for all the existing numpy functions for dtype checks seems unhealthy.

**Comment by mruberry** [comment]
And `np.iscomplex` doesn't check the dtype of the array/tensor, it's a unary elementwise operation that returns True whenever the imaginary part of its input is nonzero and False o.w.

**Comment by anjali411** [comment]
> And `np.iscomplex` doesn't check the dtype of the array/tensor, it's a unary elementwise operation that returns True whenever the imaginary part of its input is nonzero and False o.w.

hmm that doesn't seem like a semantic we'd wanna have in `PyTorch`.
